File: SymptomCheck-backend/userservice/src/main/java/com/SymptomCheck/userservice/services/AuthService.java
package com.SymptomCheck.userservice.services;

import lombok.extern.slf4j.Slf4j;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;

import java.util.Map;

@Service
@Slf4j
public class AuthService {

    private final RestTemplate restTemplate = new RestTemplate();
    private final String keycloakUrl = "http://localhost:8080";
    private final String realm = "symptomcheck-realm";
    private final String clientId = "userservice";
    private final String clientSecret = "userservice-secret";

    public Map<String, Object> login(String username, String password) {
        try {
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

            MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
            body.add("client_id", clientId);
            body.add("client_secret", clientSecret);
            body.add("username", username);
            body.add("password", password);
            body.add("grant_type", "password");
            body.add("scope", "openid");

            HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(body, headers);

            String tokenUrl = keycloakUrl + "/realms/" + realm + "/protocol/openid-connect/token";
            ResponseEntity<Map> response = restTemplate.postForEntity(tokenUrl, request, Map.class);

            if (response.getStatusCode().is2xxSuccessful()) {
                return response.getBody();
            } else {
                throw new RuntimeException("Invalid credentials");
            }

        } catch (Exception e) {
            throw new RuntimeException("Login failed: " + e.getMessage());
        }
    }

    public Map<String, Object> refreshToken(String refreshToken) {
        try {
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

            MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
            body.add("client_id", clientId);
            body.add("client_secret", clientSecret);
            body.add("grant_type", "refresh_token");
            body.add("refresh_token", refreshToken);

            HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(body, headers);

            String tokenUrl = keycloakUrl + "/realms/" + realm + "/protocol/openid-connect/token";
            ResponseEntity<Map> response = restTemplate.postForEntity(tokenUrl, request, Map.class);

            if (response.getStatusCode().is2xxSuccessful()) {
                return response.getBody();
            } else {
                throw new RuntimeException("Token refresh failed");
            }

        } catch (Exception e) {
            throw new RuntimeException("Token refresh failed: " + e.getMessage());
        }
    }

    public void logout(String refreshToken) {
        try {
            HttpHeaders headers = new HttpHeaders();
            headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

            MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
            body.add("client_id", clientId);
            body.add("client_secret", clientSecret);
            body.add("refresh_token", refreshToken);

            HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(body, headers);

            String logoutUrl = keycloakUrl + "/realms/" + realm + "/protocol/openid-connect/logout";
            restTemplate.postForEntity(logoutUrl, request, String.class);


        } catch (Exception e) {
            throw new RuntimeException("Logout failed: " + e.getMessage());
        }
    }
}

File: SymptomCheckFrontend/src/app/services/auth.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, Observable, from } from 'rxjs';
import { tap, switchMap } from 'rxjs/operators';
import { User, UserRole } from '../models/user.model';
import { UserRegistrationRequest } from '../models/auth-request.model';
import { KeycloakService } from 'keycloak-angular';
import { Router } from '@angular/router';
import { DoctorProfileStatusDTO } from '../models/doctor-profile-status.model';

const apiUrl = 'http://userservice/api/v1/users';


@Injectable({
  providedIn: 'root'
})
export class AuthService {
  private currentUserSubject = new BehaviorSubject<User | null>(null);
  public currentUser$ = this.currentUserSubject.asObservable();


  constructor(
    private http: HttpClient,
    private keycloakService: KeycloakService,
    private router: Router,

  ) {
    this.loadCurrentUser();
  }

  public loadCurrentUser() {
    // Check if user is authenticated in Keycloak
    if (this.keycloakService.isLoggedIn()) {
      this.keycloakService.getToken().then(token => {
        console.log('✅ Access Token:', token);

        // Fetch user details from backend
        this.http.get<User>(`${apiUrl}/me`).subscribe({
          next: user => {
            console.log('✅ Current User:', user);
            this.currentUserSubject.next(user);
          },
          error: err => {
            console.error('❌ Error fetching user:', err);
            this.currentUserSubject.next(null);
          }
        });
      });
    }
  }


  /**
   * Register a new user (this doesn't require Keycloak login)
   */
  registerUser(userData: any, file: File | null) {
    const formData = new FormData();
    formData.append('user', new Blob([JSON.stringify(userData)], { type: 'application/json' }));

    if (file) {
      formData.append('file', file);
    }

    return this.http.post(`${apiUrl}/register`, formData);
  }

  /**
   * Login using Keycloak (redirects to Keycloak or uses custom login)
   * Option 1: Redirect to Keycloak login page
   */
  loginWithRedirect(): void {
    this.keycloakService.login({
      redirectUri: window.location.origin + '/dashboard'
    });
  }

  /**
   * Login using custom form with Keycloak's Direct Access Grant
   * This allows you to use your custom login page while still using Keycloak tokens
   */
  async loginWithCredentials(username: string, password: string): Promise<void> {
    try {
      // Call Keycloak token endpoint directly
      const response = await fetch('http://localhost:8080/realms/symptomcheck-realm/protocol/openid-connect/token', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: new URLSearchParams({
          client_id: 'angular-client',
          grant_type: 'password',
          username: username,
          password: password,
          scope: 'openid'
        })
      });

      if (!response.ok) {
        throw new Error('Invalid credentials');
      }

      const data = await response.json();
      console.log('✅ Login successful, tokens received:', data);

      // Store tokens in Keycloak service (it handles them internally, no localStorage needed by you)
      // Keycloak service will automatically use these tokens for subsequent requests
      await this.keycloakService.init({
        config: {
          url: 'http://localhost:8080',
          realm: 'symptomcheck-realm',
          clientId: 'angular-client',
        },
        initOptions: {
          token: data.access_token,
          refreshToken: data.refresh_token,
          checkLoginIframe: false,
        },
        enableBearerInterceptor: true,
        bearerExcludedUrls: ['/assets', '/api/v1/users/register', '/api/v1/auth']
      });

      // Load current user after successful login
      this.loadCurrentUser();

      // Navigate to dashboard
      this.router.navigate(['/dashboard']);

    } catch (error) {
      console.error('❌ Login failed:', error);
      throw error;
    }
  }

  /**
   * Logout from Keycloak
   */
  logout(): void {
    this.keycloakService.logout(window.location.origin);
    this.currentUserSubject.next(null);
  }

  /**
   * Check if user is authenticated
   */
  isAuthenticated(): boolean {
    return this.keycloakService.isLoggedIn();
  }


  getUserRoles(): string[] {
  return this.keycloakService?.getUserRoles(true) || [];
}

  /**
   * Get current user
   */
  getCurrentUser(): User | null {
    return this.currentUserSubject.value;
  }

  /**
   * Get access token for manual API calls
   */
  async getAccessToken(): Promise<string> {
    return await this.keycloakService.getToken();
  }
}
